### ISSUE TITLE
Tolerate unexpected auth success/failure messages

### ISSUE THREAD

**Issue by seredvladislav** [ronf/asyncssh#221]
Hello, thank you for the very nice and **very fast** ssh-client implementation in python. I have a few very old ssh servers and they sent some strange packets after(!) authentication success. Is there any way to skip that packets? ( for example paramiko just warning). I've read documentation a few times, but didn't find any related info.
```
DEBUG:asyncio:Using selector: SelectSelector
INFO:asyncssh:Opening SSH connection to ssh_host, port 22
INFO:asyncssh:[conn=0] Connection to ssh_host, port 22 succeeded
INFO:asyncssh:[conn=0]   Local address: 10.10.0.14, port 53601
DEBUG:asyncssh:[conn=0] Requesting key exchange
DEBUG:asyncssh:[conn=0] Received key exchange request
DEBUG:asyncssh:[conn=0] Beginning key exchange
DEBUG:asyncssh:[conn=0] Completed key exchange
INFO:asyncssh:[conn=0] Beginning auth for user ssh_user
DEBUG:asyncssh:[conn=0] Trying password auth
INFO:asyncssh:[conn=0] Auth for user ssh_user succeeded
INFO:asyncssh:[conn=0] Creating local SOCKS forwarder on 127.0.0.1, port 7000
listen on port: 7000 for 127.0.0.1
INFO:asyncssh:[conn=0] Sending disconnect: Unexpected userauth response (2)
INFO:asyncssh:[conn=0] Connection failure: Unexpected userauth response
```
Code:
```
import asyncio, asyncssh, sys
import pdb
import time
import logging

logging.basicConfig(
    level=logging.DEBUG,
)


async def run_socks(ssh_host, ssh_user, ssh_password, socks_host, socks_port):
    conn, client = await asyncio.wait_for(
        asyncssh.create_connection(
            None, host=ssh_host, username=ssh_user,
            password=ssh_password, known_hosts=None, agent_path=None,
            # keepalive_interval=2
        )
        , 20)
    listener = await conn.forward_socks(socks_host, socks_port)
    print("listen on port: " + str(listener.get_port()) + " for " + str(socks_host))
    while 1:
        await asyncio.sleep(15)


def main():
    ssh_host, ssh_user, ssh_password, socks_host, socks_port = 'ssh_host', 'ssh_user', 'ssh_password', '127.0.0.1', 7000
    try:
        asyncio.get_event_loop().run_until_complete(run_socks(ssh_host, ssh_user, ssh_password, socks_host, socks_port))
    except (OSError, asyncssh.Error) as exc:
        sys.exit('SSH connection failed: ' + str(exc))


if __name__ == '__main__':
    main()
```

**Comment by ronf** [comment]
It would appear that the server is sending multiple MSG_USERAUTH_SUCCESS packets. Can you confirm that's what you are seeing here? Setting the AsyncSSH debug level to 3 should allow you to see the exact packets appearing on the connection.

That's a pretty strange thing to do if the client is only sending one MSG_USERAUTH_REQUEST, and even if there were multiple auth requests sent in parallel (which should never be the case with AsyncSSH), RFC 4252 is pretty clear on this point:

     SSH_MSG_USERAUTH_SUCCESS MUST be sent only once.  When
     SSH_MSG_USERAUTH_SUCCESS has been sent, any further authentication
     requests received after that SHOULD be silently ignored.

That said, if there are broken server implementations out there which are doing this, I could fairly easily have AsyncSSH ignore additional SSH_MSG_USERAUTH_SUCCESS messages after auth has succeeded. I just want to make sure that's what's happening before I put in such a workaround.

**Comment by seredvladislav** [comment]
Hi, Ron. I'v got MSG_USERAUTH_FAILURE packet, not multiple MSG_USERAUTH_SUCCESS. But in any case follow RFC 4252 we should just skip such packet - ".... any further authentication
 requests received after that SHOULD be silently ignored." Is it right?
```
DEBUG:asyncssh:[conn=0, pktid=8] Sent MSG_USERAUTH_REQUEST (50), 50 bytes
....
DEBUG:asyncssh:[conn=0, pktid=5] Received MSG_USERAUTH_SUCCESS (52), 1 byte
...
INFO:asyncssh:[conn=0] Creating local SOCKS forwarder on 127.0.0.1, port 7000
...
DEBUG:asyncssh:[conn=0, pktid=6] Received MSG_USERAUTH_FAILURE (51), 24 bytes
...
DEBUG:asyncssh:[conn=0] Remaining auth methods: publickey,password
INFO:asyncssh:[conn=0] Sending disconnect: Unexpected userauth response (2)
DEBUG:asyncssh:[conn=0, pktid=9] Sent MSG_IGNORE (2), 5 bytes
DEBUG:asyncssh:[conn=0, pktid=10] Sent MSG_DISCONNECT (1), 46 bytes
INFO:asyncssh:[conn=0] Connection failure: Unexpected userauth response
```

**Comment by ronf** [comment]
Huh! I definitely wasn't expecting to see a FAILURE message after the SUCCESS.

I put together a test here which confirms that OpenSSH doesn't fail if it gets any USERAUTH packets after sending USERAUTH_SUCCESS, but it does send a MSG_UNIMPLEMENTED (packet type 3) in that case, and logs something like the following in its debug ("-vvv") output:

```
debug3: receive packet: type 52
debug1: Authentication succeeded (publickey).
Authenticated to localhost ([::1]:8022).
...
debug3: receive packet: type 51
dispatch_protocol_error: type 51 seq 17
debug3: send packet: type 3
```

When you connect to one of these servers with OpenSSH with "-vvv", are you seeing the "receive packet: type 51" after the authentication succeeds? Are you seeing the "dispatch_protocol_error" and the "send packet: type 3"? I'm just wondering if there might be something else AsyncSSH is doing which is triggering the spurious FAILURE message, or if other SSH client implementations are also seeing this message but ignoring it.

The RFC is a bit unclear about whether such an unexpected response should be ignored or not. The text says "any further authentication requests received after that SHOULD be silently ignored". However, that seems to be discussing the behavior of an SSH server when it gets additional MSG_USERAUTH_REQUEST messages after auth succeeds, not what a client should do if it gets additional MSG_USERAUTH_SUCCESS or MSG_USERAUTH_FAILURE messages.

It makes sense for a server to ignore additional auth request messages after a success response, as a client might have sent multiple auth requests in parallel without waiting for the response. A server would need to ignore these later auth requests (not sending back either a SUCCESS or FAILURE response) to allow the connecting to continue. However, there's no obvious reason why a client should be expected to ignore unexpected auth response messages.

For better compatibility with OpenSSH's behavior, I'm willing to consider ignoring these messages anyway, but I'm curious if there's something else AsyncSSH is doing that's triggering these messages,. For instance, I found another place where AsyncSSH had trouble working with an Erlang SSH server and it was related to sending a MSG_IGNORE in a place where the Erlang implementation wasn't expecting one.

> P.S.
> Additionally I have one more question:
> Should we close socks5 listener if we have closed connection with ssh server or not?

That shouldn't be necessary. Any channels or forwarding actions which are active on an SSH connection will automatically be cleaned up by AsyncSSH when the connection to the server is closed.

As for the listener.wait_closed(), you can do that if you want, but it might make more sense to just do a conn.wait_closed() there. By the time that returns, the SOCKS listener and all other state associated with the server connection would have been cleaned up.

**Comment by seredvladislav** [comment]
> When you connect to one of these servers with OpenSSH with "-vvv", are you seeing the "receive packet: type 51" after the authentication succeeds? Are you seeing the "dispatch_protocol_error" and the "send packet: type 3"? 

```
debug1: Authentication succeeded (password).
....
debug1: channel 0: new [client-session]
...
debug3: receive packet: type 51 <------
dispatch_protocol_error: type 51 seq 7 <------
debug3: send packet: type 3
debug3: receive packet: type 91
debug2: callback start
......
debug3: receive packet: type 3
debug1: Received SSH2_MSG_UNIMPLEMENTED for 8 <----
debug3: receive packet: type 99
debug2: channel_input_status_confirm: type 99 id 0
debug2: PTY allocation request accepted on channel 0
debug3: receive packet: type 99
debug2: channel_input_status_confirm: type 99 id 0
debug2: shell request accepted on channel 0
```
Also have a look: https://github.com/paramiko/paramiko/blob/dccc563c3a581a14ceaa261b5fe0d08539f84b87/paramiko/transport.py#L2118

> As for the listener.wait_closed(), you can do that if you want, but it might make more sense to just do a conn.wait_closed() there. By the time that returns, the SOCKS listener and all other state associated with the server connection would have been cleaned up.      

Thank you for suggestion, I'm new be in asyncio.
P.S.
Sorry for my very short answers, it happened because your English much better than my.

**Comment by ronf** [comment]
Thanks very much for the confirmation that you are getting this unexpected message from the server even when using other clients.

Looking at AsyncSSH, it already ignores MSG_USERAUTH_REQUEST when it is received after authentication has succeeded, so there's no issue there. It also handles MSG_USERAUTH_BANNER just fine whether auth has completed or not, and it already sends MSG_UNIMPLEMENTED for any other messages in the dedicated auth range when auth is not in progress. So, the only issues are with MSG_USERAUTH_SUCCESS and MSG_USERAUTH_FAILURE, where it was explicitly raising a ProtocolError. I believe the following change should make AsyncSSH more forgiving:

```patch
diff --git a/asyncssh/connection.py b/asyncssh/connection.py
index 7b84cf8..c42d787 100644
--- a/asyncssh/connection.py
+++ b/asyncssh/connection.py
@@ -1698,7 +1698,8 @@ class SSHConnection(SSHPacketHandler):
             # pylint: disable=no-member
             self.try_next_auth()
         else:
-            raise ProtocolError('Unexpected userauth response')
+            self.logger.debug2('Unexpected userauth failure response')
+            self.send_packet(MSG_UNIMPLEMENTED, UInt32(pktid))
 
     def _process_userauth_success(self, pkttype, pktid, packet):
         """Process a user authentication success response"""
@@ -1738,7 +1739,8 @@ class SSHConnection(SSHPacketHandler):
                 self._waiter.set_result(None)
                 self._wait = None
         else:
-            raise ProtocolError('Unexpected userauth response')
+            self.logger.debug2('Unexpected userauth success response')
+            self.send_packet(MSG_UNIMPLEMENTED, UInt32(pktid))
 
     def _process_userauth_banner(self, pkttype, pktid, packet):
         """Process a user authentication banner message"""
```

I did a quick test here against a server where I forced a MSG_USERAUTH_FAILURE to be sent after MSG_USERAUTH_SUCCESS and it seemed to work. Would you mind trying this against the server where you are seeing the problem? If it looks good, I'll get this checked into the "develop" branch and roll it into the next release.

> Sorry for my very short answers, it happened because your English much better than my.

No problem at all - thanks so much for taking the time to report this and for your help in analyzing it!

**Comment by ronf** [comment]
This change is now available in the "develop" branch as commit 0ece24d. Let me know if you run into any problems with it. If needed, I can still adjust it before moving it into the master branch.

**Comment by seredvladislav** [comment]
```
pip install git+https://github.com/ronf/asyncssh.git@develop --upgrade
```
Wow! Now everything works as expected on my side and all my current problems has been solved. Thank you so much for your time.

**Comment by ronf** [comment]
Great - thank you for the report! I'll make sure this change is in the next release.

**Comment by ronf** [comment]
This change is now available in AsyncSSH 1.17.1.
